File: pagination/python/src/pagination/page_request.py
from dataclasses import dataclass, field
from typing import List


DEFAULT_PAGE_NUMBER = 1
DEFAULT_PAGE_SIZE = 10
DEFAULT_TOTAL_ITEMS = 0


@dataclass(frozen=True)
class PageRequest:
    page_number: int
    page_size: int
    total_items: int
    total_pages: int = field(init=False)
    start_item: int = field(init=False)
    end_item: int = field(init=False)
    has_previous: bool = field(init=False)
    has_next: bool = field(init=False)

    def __post_init__(self) -> None:
        if self.total_items < 0:
            raise ValueError("totalItems must be >= 0")
        if self.page_size < 1:
            raise ValueError("pageSize must be >= 1")

        total_pages = (
            0 if self.total_items == 0
            else (self.total_items + self.page_size - 1) // self.page_size
        )
        page_number = _clamp_page_number(self.page_number, total_pages)
        start_item = 0 if self.total_items == 0 else (page_number - 1) * self.page_size + 1
        end_item = (
            0 if self.total_items == 0
            else min(page_number * self.page_size, self.total_items)
        )

        object.__setattr__(self, "page_number", page_number)
        object.__setattr__(self, "total_pages", total_pages)
        object.__setattr__(self, "start_item", start_item)
        object.__setattr__(self, "end_item", end_item)
        object.__setattr__(self, "has_previous", page_number > 1)
        object.__setattr__(self, "has_next", page_number < total_pages)

    def page_window(self, window_size: int) -> List[int]:
        if window_size <= 0 or self.total_pages == 0:
            return []
        size = min(window_size, self.total_pages)
        half = size // 2
        start = self.page_number - half
        end = start + size - 1
        if start < 1:
            start = 1
            end = size
        if end > self.total_pages:
            end = self.total_pages
            start = end - size + 1
        return list(range(start, end + 1))


def _clamp_page_number(requested: int, total_pages: int) -> int:
    if total_pages == 0:
        return 1
    if requested < 1:
        return 1
    if requested > total_pages:
        return total_pages
    return requested

File: pagination/csharp/src/Pagination/PageRequest.cs
namespace Pagination;

public sealed record PageRequest
{
    public int PageNumber { get; }
    public int PageSize { get; }
    public int TotalItems { get; }

    public PageRequest(int pageNumber, int pageSize, int totalItems)
    {
        if (totalItems < 0)
            throw new ArgumentException("totalItems must be >= 0", nameof(totalItems));
        if (pageSize < 1)
            throw new ArgumentException("pageSize must be >= 1", nameof(pageSize));

        PageSize = pageSize;
        TotalItems = totalItems;
        PageNumber = ClampPageNumber(pageNumber, TotalPagesFor(totalItems, pageSize));
    }

    public int TotalPages => TotalPagesFor(TotalItems, PageSize);

    public int StartItem => TotalItems == 0 ? 0 : (PageNumber - 1) * PageSize + 1;

    public int EndItem => TotalItems == 0 ? 0 : Math.Min(PageNumber * PageSize, TotalItems);

    public bool HasPrevious => PageNumber > 1;

    public bool HasNext => PageNumber < TotalPages;

    public IReadOnlyList<int> PageWindow(int windowSize)
    {
        if (windowSize <= 0 || TotalPages == 0) return Array.Empty<int>();

        var size = Math.Min(windowSize, TotalPages);
        var half = size / 2;
        var start = PageNumber - half;
        var end = start + size - 1;

        if (start < 1) { start = 1; end = size; }
        if (end > TotalPages) { end = TotalPages; start = end - size + 1; }

        var window = new int[size];
        for (var i = 0; i < size; i++) window[i] = start + i;
        return window;
    }

    private static int TotalPagesFor(int totalItems, int pageSize) =>
        totalItems == 0 ? 0 : (totalItems + pageSize - 1) / pageSize;

    private static int ClampPageNumber(int requested, int totalPages)
    {
        if (totalPages == 0) return 1;
        if (requested < 1) return 1;
        if (requested > totalPages) return totalPages;
        return requested;
    }
}
